kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluatorSuite.scala

```diff
@@ -773,6 +773,226 @@ class DefaultExpressionEvaluatorSuite extends AnyFunSuite with ExpressionSuiteBa
     ofNull(DoubleType.DOUBLE)
   )
 
+  test("evaluate expression: substring") {
+    // scalastyle:off nonascii
+    val data = Seq[String](
+      null, "one", "two", "three", "four", null, null, "seven", "eight", "😉", "ë")
+    val col = stringVector(data)
+    val col_name = "str_col"
+    val schema = new StructType().add(col_name, StringType.STRING)
+    val input = new DefaultColumnarBatch(col.getSize, schema, Array(col))
+
+    def checkSubString(
+                   input: DefaultColumnarBatch,
+                   substringExpression: ScalarExpression,
+                   expOutputSeq: Seq[String]): Unit = {
+      val actOutputVector =
+        new DefaultExpressionEvaluator(
+          schema, substringExpression, StringType.STRING).eval(input)
+      val expOutputVector = stringVector(expOutputSeq);
+      checkStringVectors(actOutputVector, expOutputVector)
+    }
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 0),
+        // scalastyle:off nonascii
+        Seq[String](null, "one", "two", "three", "four", null, null, "seven", "eight", "😉", "ë"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 1),
+        // scalastyle:off nonascii
+        Seq[String](null, "one", "two", "three", "four", null, null, "seven", "eight", "😉", "ë"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 2),
+        Seq[String](null, "ne", "wo", "hree", "our", null, null, "even", "ight", "", "̈"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), -1),
+        // scalastyle:off nonascii
+        Seq[String](null, "e", "o", "e", "r", null, null, "n", "t", "😉", "̈"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), -1000),
+        // scalastyle:off nonascii
+        Seq[String](null, "one", "two", "three", "four", null, null, "seven", "eight", "😉", "ë"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 0, Option(4)),
+        // scalastyle:off nonascii
+        Seq[String](null, "one", "two", "thre", "four", null, null, "seve", "eigh", "😉", "ë"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 2, Option(0)),
+        Seq[String](null, "", "", "", "", null, null, "", "", "", ""))
+
+    checkSubString(
+      input,
+      substring(new Column(col_name), 1, Option(1)),
+      // scalastyle:off nonascii
+      Seq[String](null, "o", "t", "t", "f", null, null, "s", "e", "😉", "e"))
```

Comment:
the last character (e) for (ë) result same in other systems?

Also for the emoji (does it have surrogate pairs)?

Reply:
>Also for the emoji (does it have surrogate pairs)?
iiuc yes given
System.out.println("😉".length()) -> 2

the last character (e) for (ë) result same in other systems?
Verified the following using mysql
select substring("ë",1,1) ->
| substring("ë",1,1) |
| ------------------- |
| e                   |